File: Engine/Runtime/Core/Private/Numerics/Quaternion.cpp
// Copyright 2020 Aumoa.lib. All right reserved.

#include "Numerics/Quaternion.h"

#include <algorithm>
#include "IndexOutOfRangeException.h"
#include "HashHelper.h"
#include "Numerics/AxisAlignedCube.h"
#include "Numerics/Matrix4x4.h"
#include "Numerics/Vector4.h"
#include "Mathematics/Math.h"
#include "Mathematics/TAngleDef.h"

using namespace std;

Quaternion Quaternion::Identity = Quaternion(0, 0, 0, 1.0f);

Quaternion::Quaternion()
{

}

Quaternion::Quaternion(float x, float y, float z, float w)
{
	X = x;
	Y = y;
	Z = z;
	W = w;
}

Quaternion::Quaternion(const Vector3& xyz, float w)
{
	X = xyz.X;
	Y = xyz.Y;
	Z = xyz.Z;
	W = w;
}

Quaternion::Quaternion(float splat)
{
	X = splat;
	Y = splat;
	Z = splat;
	W = splat;
}

Quaternion::Quaternion(const Quaternion& copy)
{
	X = copy.X;
	Y = copy.Y;
	Z = copy.Z;
	W = copy.W;
}

bool Quaternion::Equals(const Quaternion& rh) const
{
	return (*this) == rh;
}

bool Quaternion::NearlyEquals(const Quaternion& rh, float epsilon) const
{
	return abs(X - rh.X) <= epsilon
		&& abs(Y - rh.Y) <= epsilon
		&& abs(Z - rh.Z) <= epsilon
		&& abs(W - rh.W) <= epsilon;
}

size_t Quaternion::GetHashCode() const
{
	return HashHelper::GetHashCode(X)
		 ^ HashHelper::GetHashCode(Y)
		 ^ HashHelper::GetHashCode(Z)
		 ^ HashHelper::GetHashCode(W);
}

TRefPtr<String> Quaternion::ToString() const
{
	return String::Format(L"{{Axis: {0}, Angle: {1}}}", Axis, Angle);
}

float Quaternion::GetComponentOrDefault(size_t index) const
{
	if (!Contains(index))
	{
		return 0;
	}

	const float* ptr = &X;
	return ptr[index];
}

bool Quaternion::Contains(size_t index) const
{
	return index < Count;
}

size_t Quaternion::Count_get() const
{
	return 4;
}

pair<Vector3, float> Quaternion::ToAxisAngle() const
{
	return { Axis, Angle };
}

Vector3 Quaternion::RotateVector(const Vector3& v) const
{
	auto Q = Vector3(X, Y, Z);
	Vector3 T = 2.0 * Vector3::CrossProduct(Q, v);
	Vector3 result = v + (W * T) + Vector3::CrossProduct(Q, T);

	return result;
}

Vector4 Quaternion::RotateVector(const Vector4& v) const
{
	return Vector4(RotateVector(v.Cast<Vector3>()), v.W);
}

float Quaternion::LengthSq_get() const
{
	return X * X + Y * Y + Z * Z + W * W;
}

float Quaternion::Length_get() const
{
	return sqrt(LengthSq);
}

Quaternion Quaternion::Normalized_get() const
{
	return (*this) / Length;
}

Quaternion Quaternion::Direction_get() const
{
	return Normalized;
}

void Quaternion::Direction_set(const Quaternion& value)
{
	(*this) = value * Length;
}

Vector3 Quaternion::VectorPart_get() const
{
	return Vector3(X, Y, Z);
}

void Quaternion::VectorPart_set(const Vector3& value)
{
	X = value.X;
	Y = value.Y;
	Z = value.Z;
}

float Quaternion::Angle_get() const
{
	return 2.0f * acos(W);
}

Vector3 Quaternion::Axis_get() const
{
	float S = sqrt(max(1.0f - (W * W), 0.0f));

	if (S >= 0.0001f)
	{
		return Vector3(X / S, Y / S, Z / S);
	}

	return Vector3(1.0f, 0, 0);
}

Quaternion Quaternion::Inverse_get() const
{
	return Quaternion(-X, -Y, -Z, W);
}

const float& Quaternion::operator [](size_t index) const
{
	if (!Contains(index))
	{
		throw IndexOutOfRangeException();
	}

	const float* ptr = &X;
	return ptr[index];
}

float& Quaternion::operator [](size_t index)
{
	if (!Contains(index))
	{
		throw IndexOutOfRangeException();
	}

	float* ptr = &X;
	return ptr[index];
}

Quaternion Quaternion::operator -() const
{
	return Quaternion(-X, -Y, -Z, -W);
}

Quaternion Quaternion::operator +(const Quaternion& value) const
{
	return Quaternion(X + value.X, Y + value.Y, Z + value.Z, W + value.W);
}

Quaternion Quaternion::operator -(const Quaternion& value) const
{
	return Quaternion(X - value.X, Y - value.Y, Z - value.Z, W - value.W);
}

Quaternion Quaternion::operator *(const Quaternion& value) const
{
	return Quaternion(X * value.X, Y * value.Y, Z * value.Z, W * value.W);
}

Quaternion Quaternion::operator /(const Quaternion& value) const
{
	return Quaternion(X / value.X, Y / value.Y, Z / value.Z, W / value.W);
}

float Quaternion::operator |(const Quaternion& value) const
{
	return DotProduct(*this, value);
}

Quaternion& Quaternion::operator +=(const Quaternion& right)
{
	X += right.X;
	Y += right.Y;
	Z += right.Z;
	W += right.W;
	return *this;
}

Quaternion& Quaternion::operator -=(const Quaternion& right)
{
	X -= right.X;
	Y -= right.Y;
	Z -= right.Z;
	W -= right.W;
	return *this;
}

Quaternion& Quaternion::operator *=(const Quaternion& right)
{
	X *= right.X;
	Y *= right.Y;
	Z *= right.Z;
	W *= right.W;
	return *this;
}

Quaternion& Quaternion::operator /=(const Quaternion& right)
{
	X /= right.X;
	Y /= right.Y;
	Z /= right.Z;
	W /= right.W;
	return *this;
}

float Quaternion::DotProduct(const Quaternion& left, const Quaternion& right)
{
	return left.X * right.X + left.Y * right.Y + left.Z * right.Z + left.W * right.W;
}

Quaternion Quaternion::FromAxisAngle(const Vector3& axis, TDegrees<float> angle)
{
	float half = angle.ToRadians().Value * 0.5f;
	float vsin = sin(half);
	float vcos = cos(half);

	Quaternion quat;
	quat.X = axis.X * vsin;
	quat.Y = axis.Y * vsin;
	quat.Z = axis.Z * vsin;
	quat.W = vcos;

	return quat;
}

Quaternion Quaternion::Concatenate(const Quaternion& left, const Quaternion& right)
{
	Quaternion quaternion;
	float x = (right.Y * left.Z) - (right.Z * left.Y);
	float y = (right.Z * left.X) - (right.X * left.Z);
	float z = (right.X * left.Y) - (right.Y * left.X);
	float w = ((right.X * left.X) + (right.Y * left.Y)) + (right.Z * left.Z);
	quaternion.X = ((right.X * left.W) + (left.X * right.W)) + x;
	quaternion.Y = ((right.Y * left.W) + (left.Y * right.W)) + y;
	quaternion.Z = ((right.Z * left.W) + (left.Z * right.W)) + z;
	quaternion.W = (right.W * left.W) - w;
	return quaternion;
}

Quaternion Quaternion::FromMatrix(const Matrix4x4& rotationMatrix)
{
	float side = (rotationMatrix._11 + rotationMatrix._22) + rotationMatrix._33;

	Quaternion quaternion;
	if (side > 0)
	{
		float sq = sqrt(side + 1);
		float sqiv = 0.5f / sq;
		quaternion.W = sq * 0.5f;
		quaternion.X = (rotationMatrix._23 - rotationMatrix._32) * sqiv;
		quaternion.Y = (rotationMatrix._31 - rotationMatrix._13) * sqiv;
		quaternion.Z = (rotationMatrix._12 - rotationMatrix._21) * sqiv;
	}

	else if ((rotationMatrix._11 >= rotationMatrix._22) && (rotationMatrix._11 >= rotationMatrix._33))
	{
		float sq = sqrt(((1 + rotationMatrix._11) - rotationMatrix._22) - rotationMatrix._33);
		float sqiv = 0.5f / sq;
		quaternion.X = 0.5f * sq;
		quaternion.Y = (rotationMatrix._12 + rotationMatrix._21) * sqiv;
		quaternion.Z = (rotationMatrix._13 + rotationMatrix._31) * sqiv;
		quaternion.W = (rotationMatrix._23 - rotationMatrix._32) * sqiv;
	}

	else if (rotationMatrix._22 > rotationMatrix._33)
	{
		float sq = sqrt(((1 + rotationMatrix._22) - rotationMatrix._11) - rotationMatrix._33);
		float sqiv = 0.5f / sq;
		quaternion.X = (rotationMatrix._21 + rotationMatrix._12) * sqiv;
		quaternion.Y = 0.5f * sq;
		quaternion.Z = (rotationMatrix._32 + rotationMatrix._23) * sqiv;
		quaternion.W = (rotationMatrix._31 - rotationMatrix._13) * sqiv;
	}

	else
	{
		float sq = sqrt(((1 + rotationMatrix._33) - rotationMatrix._11) - rotationMatrix._22);
		float sqiv = 0.5f / sq;
		quaternion.X = (rotationMatrix._31 + rotationMatrix._13) * sqiv;
		quaternion.Y = (rotationMatrix._32 + rotationMatrix._23) * sqiv;
		quaternion.Z = 0.5f * sq;
		quaternion.W = (rotationMatrix._12 - rotationMatrix._21) * sqiv;
	}

	return quaternion;
}

Quaternion Quaternion::FromEuler(TDegrees<float> yaw, TDegrees<float> pitch, TDegrees<float> roll)
{
	auto yawq = FromAxisAngle(Vector3::Up, yaw.ToRadians().Value);
	auto pitchq = FromAxisAngle(Vector3::Right, pitch.ToRadians().Value);
	auto rollq = FromAxisAngle(Vector3::Forward, roll.ToRadians().Value);

	return Concatenate(Concatenate(yawq, pitchq), rollq);
}

Quaternion Quaternion::Lerp(const Quaternion& left, const Quaternion& right, float t)
{
	float Dot = DotProduct(left, right);
	float Bias = Dot >= 0 ? 1.0f : -1.0f;
	return (right * t) + (left * (Bias * (1 - t)));
}

Quaternion Quaternion::Slerp(const Quaternion& left, const Quaternion& right, float t)
{
	float Threshold = 0.9995f;

	Quaternion v0 = left.Normalized;
	Quaternion v1 = right.Normalized;

	float dot = DotProduct(v0, v1);

	if (dot < 0)
	{
		v1 = -v1;
		dot = -dot;
	}

	float theta_0 = acos(dot);
	float theta = theta_0 * t;
	theta = clamp(theta, 0.0f, 3.1415926535f);
	if (theta > theta_0)
	{
		return left;
	}

	if (dot > Threshold)
	{
		return Lerp(v0, v1, t).Normalized;
	}

	float sin_theta = sin(theta);
	float sin_theta_0 = sin(theta_0);

	float s0 = cos(theta) - dot * sin_theta / sin_theta_0;
	float s1 = sin_theta / sin_theta_0;

	return v0 * s0 + v1 * s1;
}

Quaternion operator +(float left, const Quaternion& right)
{
	return Quaternion(left) + right;
}

Quaternion operator -(float left, const Quaternion& right)
{
	return Quaternion(left) - right;
}

Quaternion operator *(float left, const Quaternion& right)
{
	return Quaternion(left) * right;
}

Quaternion operator /(float left, const Quaternion& right)
{
	return Quaternion(left) / right;
}

File: Engine/Source/Runtime/Numerics/Script/VectorInterface/Quaternion.cs
using System;
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;

namespace Ayla;

[StructLayout(LayoutKind.Sequential)]
public struct Quaternion : IEquatable<Quaternion>
{
    public double X;
    public double Y;
    public double Z;
    public double W;

    public Quaternion(double x, double y, double z, double w)
    {
        X = x;
        Y = y;
        Z = z;
        W = w;
    }

    public Quaternion(in Vector3 v, double w)
    {
        X = v.X;
        Y = v.Y;
        Z = v.Z;
        W = w;
    }

    public Quaternion(in Vector4 v)
    {
        X = v.X;
        Y = v.Y;
        Z = v.Z;
        W = v.W;
    }

    public override string ToString() => $"({X}, {Y}, {Z}, {W})";
    public override int GetHashCode() => HashCode.Combine(X, Y, Z, W);
    public override bool Equals(object? obj) => obj is Quaternion other && Equals(other);

    public bool Equals(Quaternion other) => this == other;

    public bool NearlyEquals(in Quaternion other, double tolerance = 1e-8)
    {
        return Math.Abs(X - other.X) <= tolerance
            && Math.Abs(Y - other.Y) <= tolerance
            && Math.Abs(Z - other.Z) <= tolerance
            && Math.Abs(W - other.W) <= tolerance;
    }

    public double this[int index]
    {
        get => index switch
        {
            0 => X,
            1 => Y,
            2 => Z,
            3 => W,
            _ => throw new ArgumentOutOfRangeException(nameof(index), "Index must be 0, 1, 2, or 3.")
        };
        set
        {
            switch (index)
            {
                case 0: X = value; break;
                case 1: Y = value; break;
                case 2: Z = value; break;
                case 3: W = value; break;
                default: throw new ArgumentOutOfRangeException(nameof(index), "Index must be 0, 1, 2, or 3.");
            }
        }
    }

    public double LengthSq => X * X + Y * Y + Z * Z + W * W;

    public double Length => Math.Sqrt(LengthSq);

    public Quaternion Normalized => this / Length;

    public static double Dot(in Quaternion a, in Quaternion b)
        => a.X * b.X + a.Y * b.Y + a.Z * b.Z + a.W * b.W;

    public static Quaternion Conjugate(in Quaternion q)
        => new(-q.X, -q.Y, -q.Z, q.W);

    public Quaternion Conjugate() => Conjugate(this);

    public static Quaternion Inverse(in Quaternion q)
    {
        double lenSq = q.LengthSq;
        if (lenSq == 0) return Identity;
        var conj = Conjugate(q);
        return new(conj.X / lenSq, conj.Y / lenSq, conj.Z / lenSq, conj.W / lenSq);
    }

    public Quaternion Inverse() => Inverse(this);

    public static Quaternion operator -(in Quaternion q)
        => new(-q.X, -q.Y, -q.Z, -q.W);

    public static Quaternion operator +(in Quaternion a, in Quaternion b)
        => new(a.X + b.X, a.Y + b.Y, a.Z + b.Z, a.W + b.W);

    public static Quaternion operator -(in Quaternion a, in Quaternion b)
        => new(a.X - b.X, a.Y - b.Y, a.Z - b.Z, a.W - b.W);

    public static Quaternion operator *(in Quaternion a, double s)
        => new(a.X * s, a.Y * s, a.Z * s, a.W * s);

    public static Quaternion operator *(double s, in Quaternion a)
        => new(a.X * s, a.Y * s, a.Z * s, a.W * s);

    public static Vector3 operator *(in Quaternion q, in Vector3 v) => q.Transform(v);

    public static Quaternion operator /(in Quaternion a, double s)
        => new(a.X / s, a.Y / s, a.Z / s, a.W / s);

    public static Quaternion operator *(in Quaternion a, in Quaternion b)
    {
        // Quaternion multiplication (Hamilton product)
        return new(
            a.W * b.X + a.X * b.W + a.Y * b.Z - a.Z * b.Y,
            a.W * b.Y - a.X * b.Z + a.Y * b.W + a.Z * b.X,
            a.W * b.Z + a.X * b.Y - a.Y * b.X + a.Z * b.W,
            a.W * b.W - a.X * b.X - a.Y * b.Y - a.Z * b.Z
        );
    }

    public static bool operator ==(in Quaternion a, in Quaternion b)
        => a.X == b.X && a.Y == b.Y && a.Z == b.Z && a.W == b.W;

    public static bool operator !=(in Quaternion a, in Quaternion b)
        => a.X != b.X || a.Y != b.Y || a.Z != b.Z || a.W != b.W;

    public static explicit operator Vector4(in Quaternion q) => new(q.X, q.Y, q.Z, q.W);
    public static explicit operator Quaternion(in Vector4 v) => new(v);

    public static Quaternion Lerp(in Quaternion a, in Quaternion b, double t)
        => (a * (1.0 - t) + b * t).Normalized;

    public static Quaternion Slerp(in Quaternion a, Quaternion b, double t)
    {
        double dot = Dot(a, b);
        const double threshold = 0.9995;
        if (dot < 0.0)
        {
            b = -b;
            dot = -dot;
        }
        if (dot > threshold)
        {
            return Lerp(a, b, t);
        }
        dot = Math.Clamp(dot, -1.0, 1.0);
        double theta_0 = Math.Acos(dot);
        double theta = theta_0 * t;
        double sin_theta_0 = Math.Sin(theta_0);
        double sin_theta = Math.Sin(theta);
        double s0 = Math.Cos(theta) - dot * sin_theta / sin_theta_0;
        double s1 = sin_theta / sin_theta_0;
        return (a * s0 + b * s1).Normalized;
    }

    public static Quaternion FromAxisAngle(in Vector3 axis, double angleRad)
    {
        var norm = Vector3.Normalize(axis);
        double half = angleRad * 0.5;
        double s = Math.Sin(half);
        double c = Math.Cos(half);
        return new(norm.X * s, norm.Y * s, norm.Z * s, c);
    }

    public static Quaternion Max(in Quaternion a, in Quaternion b)
        => new(Math.Max(a.X, b.X), Math.Max(a.Y, b.Y), Math.Max(a.Z, b.Z), Math.Max(a.W, b.W));

    public static Quaternion Min(in Quaternion a, in Quaternion b)
        => new(Math.Min(a.X, b.X), Math.Min(a.Y, b.Y), Math.Min(a.Z, b.Z), Math.Min(a.W, b.W));

    public static Quaternion Abs(in Quaternion q)
        => new(Math.Abs(q.X), Math.Abs(q.Y), Math.Abs(q.Z), Math.Abs(q.W));

    public static bool IsZero(in Quaternion q)
        => q.X == 0 && q.Y == 0 && q.Z == 0 && q.W == 0;

    public static Quaternion Identity => new(0, 0, 0, 1);

    public Vector3 Transform(in Vector3 v)
    {
        // q * v * q^-1
        var qv = new Quaternion(v, 0);
        var result = this * qv * this.Inverse();
        return new Vector3(result.X, result.Y, result.Z);
    }
}
